assembly/systemCalls.ts

```diff
@@ -779,16 +779,87 @@ export namespace System {
+    const args = new system_calls.get_contract_metadata_arguments(contractId);
+    const encodedArgs = Protobuf.encode(args, system_calls.get_contract_metadata_arguments.encode);
+
+    const retcode = env.invokeSystemCall(system_call_ids.system_call_id.get_contract_metadata, SYSTEM_CALL_BUFFER.dataStart as u32, MAX_BUFFER_SIZE, encodedArgs.dataStart as u32, encodedArgs.byteLength, RETURN_BYTES.dataStart as u32);
+    checkErrorCode(retcode, SYSTEM_CALL_BUFFER.slice(0, RETURN_BYTES[0]));
+    const result = Protobuf.decode<system_calls.get_contract_metadata_result>(SYSTEM_CALL_BUFFER, system_calls.get_contract_metadata_result.decode, RETURN_BYTES[0]);
+    return result.value;
+  }
+
+  /**
+   * Legacy function to check authority for an account (not secure,
+   * it is recommended to use System.checkAuthority or
+   * System.checkCallContractAuthority)
+   * @param type type of authority required
+   * @param account account to check
+   * @param data data to be passed
+   * @returns bool true if the account has authority
+   * @example
+   * @deprecated
+   * ```ts
+   * System.checkAuthorityLegacy(authority.authorization_type.transaction_application, Base58.decode('1DQzuCcTKacbs9GGScRTU1Hc8BsyARTPqe));
+   * ```
+   */
+  export function checkAuthorityLegacy(type: authority.authorization_type, account: Uint8Array, data: Uint8Array | null = null): bool {
+    const args = new system_calls.check_authority_arguments(type, account, data !== null ? data : new Uint8Array(0));
+    const encodedArgs = Protobuf.encode(args, system_calls.check_authority_arguments.encode);
+
+    const retcode = env.invokeSystemCall(system_call_ids.system_call_id.check_authority, SYSTEM_CALL_BUFFER.dataStart as u32, MAX_BUFFER_SIZE, encodedArgs.dataStart as u32, encodedArgs.byteLength, RETURN_BYTES.dataStart as u32);
+    checkErrorCode(retcode, SYSTEM_CALL_BUFFER.slice(0, RETURN_BYTES[0]));
+    const result = Protobuf.decode<system_calls.check_authority_result>(SYSTEM_CALL_BUFFER, system_calls.check_authority_result.decode, RETURN_BYTES[0]);
+    return result.value;
+  }
+
+  /**
+   * Check authority for an account
+   * @param account account to check
+   * @param type type of authority required. By default it uses contract_call
+   * @param data data to be passed. By default it uses operation args
+   * @param caller contract caller. By default it calls the function to get the caller
+   * @returns bool true if the account has authority
+   * @example
+   * ```ts
+   * // check contract call authority
+   * System.checkAuthority(Base58.decode('1DQzuCcTKacbs9GGScRTU1Hc8BsyARTPqe));
+   *
+   * // if you already have the caller or args you can
+   * // pass them to save mana
+   * System.checkAuthority(
+   *   Base58.decode('1DQzuCcTKacbs9GGScRTU1Hc8BsyARTPqe),
+   *   authority.authorization_type.contract_call,
+   *   args,
+   *   caller
+   * );
+   * ```
+   */
+  export function checkAuthority(
```

Comment:
Lets put account as the second argument to make this function signature compatible with the previous version and not break the interface. That is the entire point of making this new version be name `checkAuthority`.

Reply:
I am not opposed to the change, but let's save the breaking change for a later date because everything else in this PR is an added feature.

Reply:
done